rsbuild migration

loveluthien opened the pull request:
**Description**

This PR closes #2479 and is to migrate the build tool from react-scripts to Rebuild. This migration is based on #2590. 

Migration points:
- create rsbuild related settings (done by Yu-Hsuan)
  - Added `NodePolyfill`, `glsl` plugins
  - Setup TypeScript decorators (legacy version)
  - Set TypeScript verbatimModuleSyntax to true and migrate the entire codebase
- Move `export enum` to the newly created `enums` folder (done by Yu-Hsuan)
- Give initial values to `mobx` observables
- Directly use `Jest` for unit tests
- Set up `eslint`
  - `prefer const`: force to use `const` if the variable was not reassigned
  - Use `typescript-eslint/consistent-type-imports` instead of turning on `verbatimModuleSyntax`. They might have some conflicts.
- Use `sass-embedded` instead of `sass` to support various platforms. Both packages are from the same team.
- frontend api documents
  - support the relocation of the export `enum`


**Checklist**

- check if CARTA can be built by running `npm install` and `npm run build`
- `npm run build` is successful on different platforms

For linked issues (if there are):
- [x] assignee and label added
- [x] GitHub Project estimate added

For the pull request:
- [x] reviewers and assignee added
- [x] GitHub Project estimate added
- [x] ~changelog updated~ / no changelog update needed
- [x] ~unit test added (for functions with no dependenies)~
- [x] ~API documentation added (for public variables and methods in stores)~

For dependencies:
- [x] e2e test passing / corresponding fix added / new e2e test created
- [x] ~protobuf version bumped~ / no protobuf version bumped needed
- [x] ~protobuf updated to the latest dev commit~ / no protobuf update needed
- [x] ~corresponding ICD test fix added (`BackendService` changed)~ / no ICD test fix needed (`BackendService` unchanged)
- [x] ~user manual prepared (for large new features)`

YuHsuan-Hwang committed:
move enum CatalogTextureType

YuHsuan-Hwang committed:
move telemetry enums

YuHsuan-Hwang committed:
move enum AlertType

YuHsuan-Hwang committed:
move animation enums

YuHsuan-Hwang committed:
move enum CatalogUpdateMode; remove unused enum CatalogCoordinate

YuHsuan-Hwang committed:
move enum RadiusUnits

YuHsuan-Hwang committed:
move enum DialogId

YuHsuan-Hwang committed:
move file browser enums

YuHsuan-Hwang committed:
move contour enums

YuHsuan-Hwang committed:
move region enums

YuHsuan-Hwang committed:
move enum FrameScaling

YuHsuan-Hwang committed:
move enum VectorOverlaySource

YuHsuan-Hwang committed:
move region font enums

YuHsuan-Hwang committed:
move enum CoordinateMode

YuHsuan-Hwang committed:
move enum HelpType

YuHsuan-Hwang committed:
move hips enums

YuHsuan-Hwang committed:
move enum LayoutDialogMode

YuHsuan-Hwang committed:
move overlay enums

YuHsuan-Hwang committed:
move enum PreferenceKeys

YuHsuan-Hwang committed:
move line fitting enums

YuHsuan-Hwang committed:
move catalog enums

YuHsuan-Hwang committed:
move enum LayerListSettingsTabs

YuHsuan-Hwang committed:
move enum PVAxis

YuHsuan-Hwang committed:
move region enums

YuHsuan-Hwang committed:
move line query enums

YuHsuan-Hwang committed:
move enum WidgetType

YuHsuan-Hwang committed:
remove export for ComparisonOperator

YuHsuan-Hwang committed:
move enum MultiProfileCategory

YuHsuan-Hwang committed:
move enum StokesCoordinate

YuHsuan-Hwang committed:
move enum MomentSelectingMode

YuHsuan-Hwang committed:
fix enum imports; fix typo

YuHsuan-Hwang committed:
add type; adjust path

YuHsuan-Hwang committed:
fix import web worker

YuHsuan-Hwang committed:
adjust observable initialization

YuHsuan-Hwang committed:
fix env variable

loveluthien committed:
PreferenceStore

loveluthien committed:
tsc-silent.config.js sort file names

loveluthien committed:
AnimatorStore

loveluthien committed:
no change files

loveluthien committed:
reformat animatorStore

loveluthien committed:
ImageViewConfigStore

loveluthien committed:
tsc-silent.config

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_1

loveluthien committed:
ContourStore

loveluthien committed:
RegionSetStore

loveluthien committed:
RenderConfigStore

loveluthien committed:
ColorbarStore

loveluthien committed:
RestFreqStore

loveluthien committed:
VectorOverlayConfigStore

loveluthien committed:
tsc-silent.config

YuHsuan-Hwang committed:
setup golden-layout

YuHsuan-Hwang committed:
adjust observable initialization

YuHsuan-Hwang committed:
fix production build

YuHsuan-Hwang committed:
change build folder name

YuHsuan-Hwang committed:
adjust page title

YuHsuan-Hwang committed:
remove npm run eject

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/build_tool_migration_rsbuild

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_1

loveluthien committed:
AnnotatinoAtore

loveluthien committed:
VectorOverlayStore

loveluthien committed:
RegionStore

loveluthien committed:
FrameStore: entry.name, entry.value, parseInt

loveluthien committed:
FrameStore: this.spatialTransform,  FrameStore[], frame.spatialTransform, polarization

loveluthien committed:
FrameStore: variable type in constructor

loveluthien committed:
FrameStore: previewData, spatialTransformAST, spectralReference, rasterScalingReference

loveluthien committed:
FrameStore: center(WCS), offsetCenter, (WCS), ImageViewSetting, CoordNumericalInput, wcs

loveluthien committed:
remove unused getImagePosFromWCS and getWCSFromImagePos functions

loveluthien committed:
amend SpectralProfileWidgetStore

loveluthien committed:
FrameStore: spectralLabel, this.spectralSystem

loveluthien committed:
FrameStore: getWcsSizeInArcsec, getImageXValueFromArcsec, getImageYValueFromArcsec, findChannelIndexByValue, fovSizeWCS

loveluthien committed:
FrameStore: spectralType, spectralCoordsSupported

loveluthien committed:
FrameStore: beam, zoomToSize, gl

loveluthien committed:
FrameStore: channel and contour

loveluthien committed:
FrameStore: intensityConfig, vectorOverlayData, wcsPoint2D, getImageXYValue, cursorMovementHandle

loveluthien committed:
FrameStore: initPVFrame, pixelUnitSizeArcsec, astSpectralTransform, renderedAxesNumbers, nativeSpectralCoordinate, headerRestFreq

loveluthien committed:
FrameStore: dirXNumber, dirYNumber, spectralNumber, stokesNumber, depthNumber, uvAxis, spectralAxis, channelValueBounds

loveluthien committed:
OverlayStore

loveluthien committed:
clean contour and frame store, new WCSValidation in model

loveluthien committed:
type declaration

loveluthien committed:
AppStore: spatialReference, spectralReference, rasterScalingReference

loveluthien committed:
AppStore: colorBlending

loveluthien committed:
AppStore: remove() in constructor

loveluthien committed:
AppStore: activeImage

loveluthien committed:
AppStore: activeFrame

loveluthien committed:
AppStore: this.activeFrame

loveluthien committed:
AppStore: this.hoveredFrame

loveluthien committed:
AppStore: requestPreviewPV

loveluthien committed:
AppStore: handleErrorStream, selectedRegion, saveWorkspace

loveluthien committed:
AppStore: loadWorkspace

loveluthien committed:
AppStore: loadDefaultFiles, resumeSession

loveluthien committed:
AppStore: estimatedTaskRemainingTime, activeFrameFileId

loveluthien committed:
AppStore: requestMoment, requestPV, requestPreviewPV, updateChannels

loveluthien committed:
AppStore: requestFitting

loveluthien committed:
AppStore: addFrame

loveluthien committed:
AppStore: loadConcatStokes, loadRemoteFile

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_1

loveluthien committed:
format fix

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_1' into null_check_appStore

loveluthien committed:
AppStore: set default fileId to -1, handleCatalogFilterStream, handleVectorTileStream, exportImage

loveluthien committed:
AppStore: updateHistogram, handleTileStream

loveluthien committed:
AppStore: handleRegionHistogramStream, handleSpatialProfileStream

loveluthien committed:
AppStore: handleSpectralProfileStream

loveluthien committed:
AppStore: openFile, appendFile

loveluthien committed:
AppStore: updateCatalogProfile, removeFrame, activeWorkspace

loveluthien committed:
AppStore: autorun,  addRegionsInBatch, tsc-silent.config.js

loveluthien committed:
App.tsx

loveluthien committed:
AnimatorComponent

loveluthien committed:
CatalogOverlay

loveluthien committed:
ChannelMapControl

loveluthien committed:
CursorInfo

loveluthien committed:
AboutDialog, CodeSnippetDialog

loveluthien committed:
ContourGeneratorPanel

loveluthien committed:
ContourDialogComponent

loveluthien committed:
Shared: Tables, modify catalog and spectral line query

loveluthien committed:
Shared: BiasContrastSelect, ColorMap, CoordNumericInput, PlotContainer, LinePlotSettingPanel

loveluthien committed:
Shared: LinePlotComponent

loveluthien committed:
Shared: RegionSelector, SafeNumericInput, ScatterPlot, SmoothingSettings, SpectralSetting

loveluthien committed:
Dialogs: DraggableDialog, ExternalPageDialog, FileBrowserDialog, FileListTable, ImageSave

loveluthien committed:
Dialog: FileInfoDialog, FittingDialog, layoutDialog

loveluthien committed:
Dialog: OnlineQueryDialog, PreferenceDialog

loveluthien committed:
Dialog: RegionDialogs

loveluthien committed:
Dialog: StokesDialog, TelemetryDialog, VectorOverlayDialog, WorkspaceDialog

loveluthien committed:
FileInfo, FloatingWidget, FloatingWidgetManager

loveluthien committed:
HelpDrawer, Histogram

loveluthien committed:
ImageView excluding RegionView

loveluthien committed:
ImageView: RegionView

loveluthien committed:
LayerList, Log, Menu, PvGenerator

loveluthien committed:
RegionList, RegionConfig

loveluthien committed:
SpatialProfiler, PvPreview

loveluthien committed:
SpectralProfile

loveluthien committed:
Stats, StokesAnalysis

loveluthien committed:
mobx v9  and react strict mode migration

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_1

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_1' into kuanchou/strict_null_check_2

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
fix RasterView for channel map mode

loveluthien committed:
reformat

loveluthien committed:
Merge branch 'dev' into rsbuild_migration_test

loveluthien committed:
fix import

loveluthien committed:
Merge branch 'kuanchou/mobx_reactv9_n_strict_mode_migration' into rsbuild_migration_test

loveluthien committed:
fix import

loveluthien committed:
fix package.json and import

loveluthien committed:
initialize observable outside constructor

loveluthien committed:
fix scss deprecation warning message

loveluthien committed:
give initial values for @observable and clean the constructor

loveluthien committed:
remove initial value from @observable progress

loveluthien committed:
fix spectral profile fitting

loveluthien committed:
fix default height of catalogPlot

loveluthien committed:
fix props.id in CatalogPlot and PvGenerator components

loveluthien committed:
Merge branch 'kuanchou/mobx_reactv9_n_strict_mode_migration' into rsbuild_migration_test

loveluthien committed:
restore wrongly removed reaction in CatalogWidgetStore

loveluthien committed:
move ConvertToGB enum from model to enums folder

loveluthien committed:
move HeaderTableColumnName to enums catalog and SpectralLineQuery and CatalogOverlay use the same HeaderTableColumnName

loveluthien committed:
remove unused columnIndex in updateSortRequest to fix the catalog source name sorting problem

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
Merge branch 'kuanchou/mobx_reactv9_n_strict_mode_migration' into rsbuild_migration_test

loveluthien committed:
fix env BUILD_DATE

loveluthien committed:
set optionalDependencies for rspack

loveluthien committed:
relax platform dependencies

loveluthien committed:
remove strict null check suppression

loveluthien committed:
replace react-scripts with jest

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_1

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_1' into kuanchou/strict_null_check_2

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
Merge branch 'kuanchou/mobx_reactv9_n_strict_mode_migration' into kuanchou/rsbuild_migration

loveluthien committed:
update package-lock

loveluthien committed:
Merge branch 'jest_migration' into kuanchou/rsbuild_migration

loveluthien committed:
update package-lock

loveluthien committed:
reformat

loveluthien committed:
add npm react-refresh

loveluthien committed:
remove redundant @rspack/binding in package.json and update package-lock.json

loveluthien committed:
fix eslint package

loveluthien committed:
set rsbuild eslint and blueprintjs/tslint-config

loveluthien committed:
remove @typescript-eslint/recommended and add prefer-const in rules

loveluthien committed:
fix additional eslint style: const to replace unresigned let and unused var

loveluthien committed:
reformat

loveluthien committed:
6.0.0-beta.1 versioning

loveluthien committed:
revert versioning 6.0.0-beta.1 and add supporting for unreleased dev

loveluthien committed:
fix pv generator for the case of spectralFrame is null

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_1' into kuanchou/strict_null_check_2

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
Merge branch 'kuanchou/mobx_reactv9_n_strict_mode_migration' into kuanchou/rsbuild_migration

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_1

loveluthien committed:
fix merge conflict

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_1' into kuanchou/strict_null_check_2

loveluthien committed:
solve merging conflict

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
Merge branch 'kuanchou/mobx_reactv9_n_strict_mode_migration' into kuanchou/rsbuild_migration

loveluthien committed:
use typescript-eslint/consistent-type-imports instead of verbatimModuleSyntax = true

loveluthien committed:
update package-lock.json

loveluthien committed:
replace sass with sass-embedded to fix build issue

loveluthien committed:
move sass-embedded to optionalDependencies

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_2

loveluthien committed:
solve merge conflicts

loveluthien committed:
fix error after merging dev

loveluthien committed:
fix lacking "specsys" can't plot spectrum bug and reformat

loveluthien committed:
correct variable naming

loveluthien committed:
refine code

loveluthien committed:
refine code

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_2

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
reformat

loveluthien committed:
minor modification

loveluthien committed:
add back @computed if possible

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_2

loveluthien committed:
reformat

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
address command and minor modifications

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
solve merge conflict

loveluthien committed:
minor modification and addressing comments

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
Merge branch 'kuanchou/mobx_reactv9_n_strict_mode_migration' into kuanchou/rsbuild_migration

loveluthien committed:
change the order of condition fileBrowserStore.fileList?.directory

loveluthien committed:
fix enums and eslint errors

loveluthien committed:
fix unit test

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_2

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
Merge branch 'kuanchou/mobx_reactv9_n_strict_mode_migration' into kuanchou/rsbuild_migration

loveluthien committed:
fix unit test

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_2

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
Merge branch 'kuanchou/mobx_reactv9_n_strict_mode_migration' into kuanchou/rsbuild_migration

loveluthien committed:
fix docusaurus.config.js

veggiesaurus reviewed (approved):
these changes generally look sensible to me

loveluthien committed:
fix as any problem and other minor type define problem

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_2

loveluthien committed:
update protobuf

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' of github.com:CARTAvis/carta-frontend into kuanchou/strict_null_check_2

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
Merge branch 'dev' into kuanchou/strict_null_check_2

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
type define

loveluthien committed:
Merge branch 'kuanchou/strict_null_check_2' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
minor modification

loveluthien committed:
Merge branch 'dev' into kuanchou/mobx_reactv9_n_strict_mode_migration

loveluthien committed:
fix console mobx warning messages

loveluthien committed:
add disposers for regionView

loveluthien committed:
update protobuf

loveluthien committed:
Merge branch 'kuanchou/mobx_reactv9_n_strict_mode_migration' into kuanchou/rsbuild_migration

loveluthien committed:
package update

loveluthien committed:
reformat

loveluthien committed:
Merge branch 'dev' into kuanchou/rsbuild_migration

loveluthien committed:
fix for e2e

loveluthien committed:
fix __dirname warning message

loveluthien committed:
fix default height

kswang1029 reviewed (approved):
looks fine on macOS and Ubuntu.

loveluthien committed:
modify settings

loveluthien committed:
merge Zhenkai/review rsbuild migration (#2680)

* update start, build-ts, and preview scripts to use jiti config loader

* enable eslint for both dev and build and update worker-loader alias

* refactor(stores): replace undefined as any with explicit null-safe types

* refactor(types): remove definite assignment assertions (!)

* build(rsbuild): use custom HTML template and fix template vars

- configure Rsbuild to use `public/index.html` via `html.template`
- replace invalid `ASSETPREFIX` with `assetPrefix` in HTML template
- bind `<title>` to `htmlWebpackPlugin.options.title`
- prefix loading logo path with `assetPrefix`
- disable dev lazy compilation (`dev.lazyCompilation: false`)
- enable auto-open browser on dev start (`server.open: true`)

* fix(rsbuild): reorder server configuration

* fix(env): migrate REACT_APP runtime flags to Rsbuild PUBLIC_ vars

- rename telemetry env keys in `.env.development` and `.env.production`:
  `REACT_APP_SKIP_TELEMETRY` -> `PUBLIC_REACT_APP_SKIP_TELEMETRY`
- update telemetry gating checks to use `process.env.PUBLIC_REACT_APP_SKIP_TELEMETRY`
  in `PreferenceDialogComponent` and `TelemetryService`
- update platform target checks to use `process.env.PUBLIC_REACT_APP_TARGET`
  in `AppStore` and `RootMenuComponent`

This aligns env access with Rsbuild’s `PUBLIC_` exposure rules so flags are defined at runtime.

* chore(deps): remove deprecated Blueprint TSLint config

- remove `@blueprintjs/tslint-config` from `devDependencies`
- update lockfile to drop TSLint-related transitive packages (`tslint`, `tslint-react`, `tsutils`, etc.)
- include npm lockfile normalization/optional platform dependency metadata updates

* fix(index.html): update title to a static value "CARTA"

* Revert "refactor(types): remove definite assignment assertions (!)"

This reverts commit cfcdd1b7ea750496a350aa56696f2975d0d24c1f.

* Revert "refactor(stores): replace undefined as any with explicit null-safe types"

This reverts commit f54d00558679e34c7fcce9bf3c1237964e05188c.

---------

Co-authored-by: Zhen-Kai Gao <zkgao@asiaa.sinica.edu.tw>

loveluthien committed:
update package-lock.json
